core/src/main/java/org/infinispan/transaction/impl/AbstractCacheTransaction.java

```diff
@@ -211,33 +212,34 @@ public int getTopologyId() {
 
    @Override
    public void addBackupLockForKey(Object key) {
-      // we need to synchronize this collection to be able to get a valid snapshot from another thread during state transfer
-      if (backupKeyLocks == null) backupKeyLocks = Collections.synchronizedSet(new HashSet<>(INITIAL_LOCK_CAPACITY));
-      backupKeyLocks.add(key);
+      // we need a synchronized collection to be able to get a valid snapshot from another thread during state transfer
+      final Set<Object> keys = backupKeyLocks.updateAndGet((value) -> value == null ? Collections.synchronizedSet(new HashSet<>(INITIAL_LOCK_CAPACITY)) : value);
+      keys.add(key);
    }
 
    public void registerLockedKey(Object key) {
-      // we need to synchronize this collection to be able to get a valid snapshot from another thread during state transfer
-      if (lockedKeys == null) lockedKeys = Collections.synchronizedSet(CollectionFactory.makeSet(INITIAL_LOCK_CAPACITY, keyEquivalence));
+      // we need a synchronized collection to be able to get a valid snapshot from another thread during state transfer
+      final Set<Object> keys = lockedKeys.updateAndGet((value) -> value == null ? Collections.synchronizedSet(CollectionFactory.makeSet(INITIAL_LOCK_CAPACITY, keyEquivalence)) : value);
       if (trace) log.tracef("Registering locked key: %s", toStr(key));
-      lockedKeys.add(key);
+      keys.add(key);
    }
 
    @Override
    public Set<Object> getLockedKeys() {
-      return lockedKeys == null ? InfinispanCollections.emptySet() : lockedKeys;
+      final Set<Object> keys = lockedKeys.get();
+      return keys == null ? InfinispanCollections.emptySet() : keys;
```

Comment:
I believe making a copy of the `lockedKeys` reference in a local variable would have been enough.